storage/imx6ull-flash/imx6ull-flash-test.c

```diff
@@ -112,6 +114,164 @@ void print_block(const uint8_t *data, unsigned int len)
 }
 
 
+/* Flips n bits in a raw block (assume n <= end - start) */
+int flip_bits(unsigned int n, unsigned int nblock, unsigned int start, unsigned int end)
+{
+	unsigned int i, j, k, offs, paddr = nblock * BLOCK_PAGES_CNT;
+	flashdrv_dma_t *dma;
+	flashdrv_meta_t *aux;
+	uint8_t **data;
+	int err = EOK;
+
+	if ((dma = flashdrv_dmanew()) == MAP_FAILED)
+		return -ENOMEM;
+
+	if ((data = malloc(BLOCK_PAGES_CNT * sizeof(uint8_t *))) == NULL) {
+		flashdrv_dmadestroy(dma);
+		return -ENOMEM;
+	}
+
+	do {
+		/* Read raw block data into memory and flip bits */
+		for (i = 0; i < BLOCK_PAGES_CNT; i++) {
+			if ((data[i] = mmap(NULL, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0)) == MAP_FAILED) {
+				err = -ENOMEM;
+				printf("mmap() failed: %d\n", err);
+				break;
+			}
+
+			if ((err = flashdrv_readraw(dma, paddr + i, data[i], FLASHDRV_PAGESZ)) < 0) {
+				printf("readraw() failed: %d\n", err);
+				break;
+			}
+
+			for (offs = i * FLASHDRV_PAGESZ; start < min(end, offs + FLASHDRV_PAGESZ); start += (end - start) / n--)
+				data[i][start - offs] ^= 1 << rand() % 8;
+		}
+
+		if (err)
+			break;
+
+		/* Erase the block */
+		if ((err = flashdrv_erase(dma, paddr)) < 0) {
+			printf("erase() failed: %d\n", err);
+			break;
+		}
+
+		/* Write back raw block data with flipped bits */
+		for (j = 0; j < BLOCK_PAGES_CNT; j++) {
+			if ((err = flashdrv_writeraw(dma, paddr + j, data[j], FLASHDRV_PAGESZ)) < 0) {
+				printf("writeraw() failed: %d\n", err);
+				break;
+			}
+
+			/* Check ECC errors */
+			aux = (flashdrv_meta_t *)(data[j] + flashinfo.writesz);
+			if ((err = flashdrv_read(dma, paddr + j, data[j], aux)) < 0) {
+				printf("read() failed: %d\n", err);
+				break;
+			}
+
+			for (k = 0; k < sizeof(aux->errors); k++) {
+				if (aux->errors[k] && (aux->errors[k] != 0xff)) {
+					printf("[%4u]: ", paddr + j);
+
+					if (aux->errors[k] == 0xfe)
+						printf("uncorrectable data in chunk %u\n", k);
+					else
+						printf("%u corrections in chunk %u\n", aux->errors[k], k);
+				}
+			}
+		}
+	} while (0);
```

Comment:
I really don't like the `do {} while(0)` syntax as it unnaturally increases the code complexity. Probably discussion with PP for another time.

Reply:
I prefer to use `goto` but it's forbidden :(